### ISSUE TITLE
Threads blocked while using python API

### ISSUE THREAD

**Issue by abubakarmuqaddas** [sysrepo/sysrepo#1212]
Hi,

I am trying to reproduce the `oven.c` code for the sysrepo-netopeer2 in python with [oven.txt](https://github.com/sysrepo/sysrepo/files/2169806/oven.txt). I follow the tutorial [here](http://www.sysrepo.org/static/doc/html/start_page.html), specifically in the trying it out section. After I edit the running config in step 3, the `oven_thread` does not output anything, and seems to have been blocked by maybe the `sr.global_loop()`. The thread does not execute unless I query the state using:

```
> get --filter-xpath /oven:*
DATA
<oven xmlns="urn:sysrepo:oven">
  <turned-on>true</turned-on>
  <temperature>200</temperature>
</oven>
<oven-state xmlns="urn:sysrepo:oven">
  <temperature>25</temperature>
  <food-inside>false</food-inside>
</oven-state>
```

After I repeatedly execute `get --filter-xpath /oven:*`, then the temperature changes and I get the notification:

```
> get --filter-xpath /oven:*
DATA
<oven xmlns="urn:sysrepo:oven">
  <turned-on>true</turned-on>
  <temperature>200</temperature>
</oven>
<oven-state xmlns="urn:sysrepo:oven">
  <temperature>200</temperature>
  <food-inside>false</food-inside>
</oven-state>

>
notification (2018-07-06T09:18:47Z)
<oven-ready xmlns="urn:sysrepo:oven"/>
```

Can you suggest how to avoid this issue?

Regards
Abubakar

**Comment by mislavn** [comment]
This is actually a known issue with SWIG Python bindings, by default they don't have multi threaded support.

I'll need to investigate this further.

**Comment by abubakarmuqaddas** [comment]
Thanks! Kindly inform if a separate process is created to run the oven thread, should it work?

**Comment by mislavn** [comment]
Sorry I'm not sure. The SWIG documentation on this subject is a little scarce. I intend to work on this issue in about a month.

**Comment by abubakarmuqaddas** [comment]
Ok thanks. I will see if I can create multiple processes to solve this issue.

**Comment by abubakarmuqaddas** [comment]
Hi Mislav,

Were you able to investigate into this?

Thanks
Abubakar

**Comment by mislavn** [comment]
Hi Abubakar,

I'm still not able to fix the problem in the Python bindings but you can avoid the issue if you create a new session in the callback.

```
        conn = sr.Connection("test-daemon")
        session = sr.Session(conn)
        values = sr.Vals(0)
        session.event_notif_send("/oven:oven-ready", values, sr.SR_EV_NOTIF_DEFAULT)
```

I have modified your previous python code, it works now. You also need to use the flag `SR_SUBSCR_CTX_REUSE` when you reuse the session like in the oven example. The full example is:

```
import sysrepo as sr
import sys

# Helper function for printing changes given operation, old and new value.

FLUSH = sys.stdout.flush


def print_change(op, old_val, new_val):
    if op == sr.SR_OP_CREATED:
        print("CREATED:")
        print(new_val.to_string(), end="")
    elif op == sr.SR_OP_DELETED:
        print("DELETED:")
        print(old_val.to_string(), end="")
    elif op == sr.SR_OP_MODIFIED:
        print("MODIFIED:")
        print("old value", end="")
        print(old_val.to_string(), end="")
        print("new value", end="")
        print(new_val.to_string(), end="")
    elif op == sr.SR_OP_MOVED:
        print("MOVED:\n", new_val.get_xpath(), " after ", old_val.get_xpath())
    print("")


# Helper function for printing events.
def ev_to_str(ev):
    if ev == sr.SR_EV_VERIFY:
        return "verify"
    elif ev == sr.SR_EV_APPLY:
        return "apply"
    elif ev == sr.SR_EV_ABORT:
        return "abort"
    else:
        return "abort"


# Function to print current configuration state.
# It does so by loading all the items of a session and printing them out.
def print_current_config(session, module_name):
    select_xpath = "/" + module_name + ":*//*"

    values = session.get_items(select_xpath)

    for i in range(values.val_cnt()):
        print(values.val(i).to_string(), end="")


def module_change_cb(sess, module_name, event, private_ctx):
    print("\n\n ========== CONFIG HAS CHANGED, CURRENT RUNNING CONFIG: ==========\n")

    try:
        print("\n\n ========== Notification " + ev_to_str(event) + " =============================================\n")
        if sr.SR_EV_APPLY == event:
            print("\n ========== CONFIG HAS CHANGED, CURRENT RUNNING CONFIG: ==========\n")
            print_current_config(sess, module_name)

        print("\n ========== CHANGES: =============================================\n")

        change_path = "/" + module_name + ":*"

        it = sess.get_changes_iter(change_path)

        while True:
            change = sess.get_change_next(it)
            if change == None:
                break
            print_change(change.oper(), change.old_val(), change.new_val())

        print("\n\n ========== END OF CHANGES =======================================\n")

        conn = sr.Connection("test-daemon")
        session = sr.Session(conn)
        values = sr.Vals(0)
        session.event_notif_send("/oven:oven-ready", values, sr.SR_EV_NOTIF_DEFAULT)

    except Exception as e:
        print(e)

    return sr.SR_ERR_OK

def test_rpc_cb(xpath, in_vals, out_vals, private_ctx):
    print("\n\n ========== RPC CALLED for removing food==========")

def oven_state_cb(xpath, holder, id, private_ctx):
    print("\n\n ========== OVEN State queried ==========")
    values = holder.allocate(2)
    values.val(0).set('/oven:oven-state/temperature', 3, sr.SR_UINT8_T)
    values.val(1).set("/oven:oven-state/food-inside", True, sr.SR_BOOL_T)
    return sr.SR_ERR_OK


if __name__ == '__main__':
    conn = sr.Connection("test-daemon")
    session = sr.Session(conn)
    subscribe = sr.Subscribe(session)

    subscribe.module_change_subscribe("oven", module_change_cb)

    subscribe.rpc_subscribe("/oven:remove-food", test_rpc_cb, None, sr.SR_SUBSCR_CTX_REUSE)

    subscribe.dp_get_items_subscribe('/oven:oven-state', oven_state_cb, None, sr.SR_SUBSCR_CTX_REUSE)

    sr.global_loop()

```

**Comment by Akijdarvajkar** [comment]
Hi,
i copy-past the above oven code but 
`subscribe.dp_get_items_subscribe('/oven:oven-state', oven_state_cb, None, sr.SR_SUBSCR_CTX_REUSE) `not able to call callback function oven_state_cb.

Thanks,
Akkij

**Comment by RPML** [comment]
Hi,

I was trying to use subscribe callback in one thread and in another dummy thread I was writing some text into a file. 
But here it gets blocked at global_loop and the other dummy thread will not be proceeding.
sr.global_loop() is blocking other threads. Could you please comment on this @mislavn 

Code:
cat SubscribeCallbackInThread.py
#! /usr/bin/python3
from _thread import start_new_thread
from time import sleep
import sysrepo as sr

def test_rpc_cb(xpath, in_vals, holder, private_ctx):
    try:
        print (" ========== RPC CALLED ==========")

        if (in_vals == None):
            return

        for n in range(in_vals.val_cnt()):
            key, value = in_vals.val(n).to_string().rsplit("=")
            print("key =%s"%key)
            print("value =%s"%value)

    except Exception as e:
        print (e)

def SysrepoDummyThread():
    print("inside SysrepoDummyThread")
    conn = sr.Connection("example_application")
    sess = sr.Session(conn)
    subscribe = sr.Subscribe(sess)
    print ("\n ========== SUBSCRIBE TO RPC CALL ==========")
    subscribe.rpc_subscribe("/oven:remove-food", test_rpc_cb, sr.SR_SUBSCR_CTX_REUSE)
    print ("before global loop in SysrepoDummyThread")
    sr.global_loop();
    print("returning from SysrepoDummyThread")

def DummyThread():
    print("inside DummyThread")
    f= open("DummyThread.txt","w+")
    import os
    i=0
    while True:
        f.write("This is line %d\r\n" %(i))
        i+=1
        f.flush()
        os.fsync(f)
        sleep(1)
        print("writing in file")
    print("returning from DummyThread")

def main():
    start_new_thread(SysrepoDummyThread,())
    start_new_thread(DummyThread,())
    print ("waiting in main for 500sec")
    sleep(500)

if __name__ == "__main__":
    main()

**Comment by adammahm** [comment]
I've been facing the same issue. Any updates on the soln?

**Comment by michalvasko** [comment]
Hi,
I do not know much about threads and Python, but I thought the problem is solved/understood in the issue #1506.

Regards,
Michal
